src/Database/Redis/Schema/RemoteJob.hs

```diff
@@ -0,0 +1,356 @@
+import Control.Monad.Catch
+import Control.Exception ( SomeAsyncException )
+
+import GHC.Generics ( Generic )
+
+
+-- | Errors that can occur in the remote job running.
+data RemoteJobError
+  = RemoteJobException String
+  | NoActiveWorkers
+  | Timeout
+  deriving ( Show, Generic )
+instance Binary RemoteJobError
+
+instance Serializable RemoteJobError where
+  fromBS = readBinary
+  toBS = showBinary
+
+-- | Identifier for the worker process, is only used for inspecting the queue
+newtype WorkerId = WorkerId { unWorkerId :: Text }
+  deriving newtype ( Show, Eq, Ord, IsString, Serializable, Binary )
+
+-- | Handle of the worker process, which can be used in the graceful shutdown procedure
+data WorkerHandle = WorkerHandle (MVar ()) ThreadId
+
+class JobQueue jq where
+  -- | The remote job protocol, a list of 'i -> o' entries indicating
+  --   this queue can contains jobs taking type 'i' as input and
+  --   returning type 'o'. Both 'i' and 'o' must have a binary instance.
+  type RPC jq :: [Type]
+
+  -- | Prefix for the Redis keys
+  keyPrefix :: BS.ByteString
+
+  -- | Which Redis instance the queue lives in.
+  type RedisInstance jq :: Instance
+  type RedisInstance jq = DefaultInstance
+
+-- | This queue contains many requests.
+-- There is only one request queue and it's read by all workers.
+data RequestQueue jq = RequestQueue
+instance JobQueue jq => Ref (RequestQueue jq) where
+  type RefInstance (RequestQueue jq) = RedisInstance jq
+  type ValueType (RequestQueue jq) = [(Priority, (UUID, Int, BSL.ByteString))]
+  toIdentifier RequestQueue =
+    colonSep [keyPrefix @jq, "requests"]
+
+-- | This set contains the requests that are currently being processed.
+data RunningJobs jq = RunningJobs
+instance JobQueue jq => Ref (RunningJobs jq) where
+  type RefInstance (RunningJobs jq) = RedisInstance jq
+  type ValueType (RunningJobs jq) = Set.Set (WorkerId, (UUID, Int, BSL.ByteString))
+  toIdentifier RunningJobs =
+    colonSep [keyPrefix @jq, "running"]
+
+-- | A box that contains only one response.
+-- For every response, a unique box is created (tagged with job ID).
+newtype ResultBox jq = ResultBox UUID
+instance JobQueue jq => Ref (ResultBox jq) where
+  type RefInstance (ResultBox jq) = RedisInstance jq
+  type ValueType (ResultBox jq) = [Either RemoteJobError BSL.ByteString]
+  toIdentifier (ResultBox uuid) =
+    colonSep [keyPrefix @jq, "result", toBS uuid]
+
+-- | A registry of all active workers
+data Workers jq = Workers
+instance JobQueue jq => Ref (Workers jq) where
+  type RefInstance (Workers jq) = RedisInstance jq
+  type ValueType (Workers jq) = [(Priority, WorkerId)]
+  toIdentifier Workers =
+    Redis.colonSep [keyPrefix @jq, "workers"]
+
+-- | Type class to check where in the 'RPC' list a i->o job occurs, which
+--   is then used together with 'CanHandle' to use the right handler.
+class FindHandler (i :: Type) (o :: Type) (xs :: [Type]) where
+  handlerIdx :: Proxy (i -> o) -> Proxy xs -> Int
```

Comment:
Perhaps `HasHandler` would be an a tiny bit neater name, in line with other typeclasses named as predicates rather than verbs?

Reply:
Done